### PR TITLE
Remove owner type and and owner_id from repo_content_units

### PR DESCRIPTION
**Pull request by barnabycourt**
https://pulp.plan.io/issues/849

### CODE REVIEW

**Comment by barnabycourt** [comment]
ok test

**Comment by barnabycourt** [comment]
ok test

**Comment by barnabycourt** [comment]
ok test

**Comment by beav** [comment]
ok test

the builder appears to be doing better today :palm_tree:

**Comment by ipanova** [comment]
@barnabycourt  was not it supposed to be this issue? https://pulp.plan.io/issues/849

**Comment by barnabycourt** [comment]
@ipanova you are absolutely correct. I have updated the link. Thanks!

**Comment by barnabycourt** [comment]
ok test

**Comment by barnabycourt** [comment]
ok test

**Comment by barnabycourt** [comment]
ok test

File: nodes/test/unit/test_conduit.py [hunk 1]
```diff
@@ -9,7 +9,6 @@
 # have received a copy of GPLv2 along with this software; if not, see
 # http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
 
-from mock import Mock
```
Comment (beav) [inline nodes/test/unit/test_conduit.py]: this is just for flake8?

File: server/pulp/server/db/migrations/0016_remove_repo_content_unit_owner_type_and_id.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    index_info = collection.index_information()
+    if "repo_id_-1_unit_type_id_-1_unit_id_-1_owner_type_-1_owner_id_-1" in index_info:
+        _logger.info("Dropping the uniqueness index that included the owner_type & owner_id")
+        collection.drop_index("repo_id_-1_unit_type_id_-1_unit_id_-1_owner_type_-1_owner_id_-1")
+
+
+def remove_duplicates(collection):
+    """
+    Remove entries from the repo_content_units collection that would be duplicates
+    after teh owner_type and owner_id fields are removed. Always remove older
```
Comment (beav) [inline server/pulp/server/db/migrations/0016_remove_repo_content_unit_owner_type_and_id.py]: s/teh/the/

File: server/pulp/server/db/migrations/0016_remove_repo_content_unit_owner_type_and_id.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    entries so that incremental publishing will not be affected
+
+    :param collection: a reference to the repo_content_units collection
+    :type collection: pymongo.collection.collection
+    """
+    units_to_remove = []
+    last_unit = {"repo_id": None, "unit_type_id": None, "unit_id": None}
+    for unit in collection.find(sort=[('repo_id', 1), ('unit_type_id', 1),
+                                      ('unit_id', 1), ('updated', 1)]):
```
Comment (beav) [inline server/pulp/server/db/migrations/0016_remove_repo_content_unit_owner_type_and_id.py]: Would it be possible to do an aggregate on all the units to find units with count > 1? That would avoid this loop

File: server/pulp/server/managers/repo/sync.py [hunk 4]
```diff
@@ -79,8 +79,7 @@ def sync(repo_id, sync_config_override=None):
         repo_importer = importer_manager.get_importer(repo_id)
 
         # Assemble the data needed for the sync
-        conduit = RepoSyncConduit(repo_id, repo_importer['id'], RepoContentUnit.OWNER_TYPE_IMPORTER,
-                                  repo_importer['id'])
+        conduit = RepoSyncConduit(repo_id, repo_importer['id'])
 
         call_config = PluginCallConfiguration(importer_config, repo_importer['config'],
                                               sync_config_override)
```
Comment (beav) [inline server/pulp/server/managers/repo/sync.py]: marking where I left off for lunch :smile:  will continue when I get back

:hamburger:

File: nodes/test/unit/test_conduit.py (hunk 1 above)
Comment (barnabycourt) [inline nodes/test/unit/test_conduit.py]: yup

File: server/test/unit/server/managers/repo/test_unit_association_query.py [hunk 5]
```diff
@@ -673,385 +636,3 @@ def _assert_unit_integrity(self, unit):
         self.assertTrue(unit['metadata']['md_1'] is not None)
         self.assertTrue(unit['metadata']['md_2'] is not None)
         self.assertTrue(unit['metadata']['md_3'] is not None)
-
-
-class GetUnitsStressTest(base.PulpServerTests):
```
Comment (beav) [inline server/test/unit/server/managers/repo/test_unit_association_query.py]: I didn't realize this was here before, good catch.

The git log indicates this was updated in 2011-ish, aside from refactors and PEP8 updates.

**Comment by beav** [comment]
The code change itself looks good. Two items:

* The migration might be able to be redone using an aggregation query instead of a loop.
* This change also needs a release note; the 2.7 notes are added as part of #1780 so you may need to add it in a new PR if this gets merged first.
* What happens if the migration is not run? Will the owner fields just be ignored on the mongo documents?

**Comment by barnabycourt** [comment]
If the migration is not run two things occur. 1) The extra data would remain in the database, this would be messy and could result in errors when we stop filtering for duplicates in the unit association query. 2) The uniqueness constraint would not be updated for the collection so it would not be enforced properly in the future.

**Comment by barnabycourt** [comment]
ok test

**Comment by barnabycourt** [comment]
ok test

File: docs/user-guide/release-notes/2.7.x.rst [hunk 6]
```diff
@@ -57,6 +57,8 @@ Binding API Changes
 
 Plugin API Changes
 ------------------
+* The undocumented fields 'owner_type' and 'owner_id' have been removed from the
+  'repo_content_units' collection.
```
Comment (beav) [inline docs/user-guide/release-notes/2.7.x.rst]: cool, thanks :+1:

**Comment by beav** [comment]
LGTM, nice work

**Pull request merged**